go/ql/lib/semmle/go/dataflow/ExternalFlow.qll

```diff
@@ -87,6 +90,93 @@ private import internal.FlowSummaryImpl::Private
 private import internal.FlowSummaryImpl::Private::External
 private import codeql.mad.ModelValidation as SharedModelVal
 
+/** Gets the prefix for a group of packages. */
+string groupPrefix() { result = "group:" }
+
+/** Gets a group that `package` is in, according to `packageGrouping`. */
+private string getGroup(string package) {
+  exists(string group |
+    FlowExtensions::packageGrouping(group, package) and
+    result = groupPrefix() + group
+  )
+}
+
+/**
+ * Holds if a source model exists for the given parameters.
+ *
+ * Note that we consider all packages in the same group.
+ */
+predicate sourceModel(
+  string package, string type, boolean subtypes, string name, string signature, string ext,
+  string output, string kind, string provenance, QlBuiltins::ExtensionId madId
+) {
+  exists(string p |
+    FlowExtensions::sourceModel(p, type, subtypes, name, signature, ext, output, kind, provenance,
+      madId)
+  |
+    not exists(string s | p = groupPrefix() + s) and package = p
+    or
+    // Also look for models that are defined for a group that `package` is part of.
+    p = getGroup(package)
+  )
+}
+
+/**
+ * Holds if a sink model exists for the given parameters.
+ *
+ * Note that we consider all packages in the same group.
+ */
+predicate sinkModel(
+  string package, string type, boolean subtypes, string name, string signature, string ext,
+  string input, string kind, string provenance, QlBuiltins::ExtensionId madId
+) {
+  exists(string p |
+    FlowExtensions::sinkModel(p, type, subtypes, name, signature, ext, input, kind, provenance,
+      madId)
+  |
+    not exists(string s | p = groupPrefix() + s) and package = p
+    or
+    // Also look for models that are defined for a group that `package` is part of.
+    p = getGroup(package)
+  )
+}
+
+/**
+ * Holds if a summary model exists for the given parameters.
+ *
+ * Note that we consider all packages in the same group.
+ */
+predicate summaryModel(
+  string package, string type, boolean subtypes, string name, string signature, string ext,
+  string input, string output, string kind, string provenance, QlBuiltins::ExtensionId madId
+) {
+  exists(string p |
+    FlowExtensions::summaryModel(p, type, subtypes, name, signature, ext, input, output, kind,
+      provenance, madId)
+  |
+    not exists(string s | p = groupPrefix() + s) and package = p
```

Comment:
Just `not p.matches(groupPrefix() + "%")`?

Reply:
It's very hard to know what's faster. I remember now that QL strings have a `prefix()` method, which surely must be performant, so I'll use that.

Reply:
The `packageOrGroup.prefix(groupPrefix().length()) = groupPrefix()` should actually be expected to be worse than both of those other options, since it forces the evaluator to manifest all those `packageOrGroup`-prefixes in the string pool. The `matches`-version is fine, but I expect `exists(string s | p = groupPrefix() + s)` to be the best version. We really need a builtin `string.startsWith` predicate to make it such that the nicest version is also the best version for performance.